sway/input/seatop_move_tiling.c

```diff
@@ -24,28 +23,17 @@ struct seatop_move_tiling_event {
 	struct sway_container *con;
 	struct sway_node *target_node;
 	enum wlr_edges target_edge;
-	struct wlr_box drop_box;
 	double ref_lx, ref_ly; // cursor's x/y at start of op
 	bool threshold_reached;
 	bool split_target;
 	bool insert_after_target;
+	struct wlr_scene_rect *indicator_rect;
 };
 
-static void handle_render(struct sway_seat *seat, struct render_context *ctx) {
+static void handle_end(struct sway_seat *seat) {
 	struct seatop_move_tiling_event *e = seat->seatop_data;
-	if (!e->threshold_reached) {
-		return;
-	}
-	if (e->target_node && node_get_output(e->target_node) == ctx->output) {
-		float color[4];
-		memcpy(&color, config->border_colors.focused.indicator,
-				sizeof(float) * 4);
-		premultiply_alpha(color, 0.5);
-		struct wlr_box box;
-		memcpy(&box, &e->drop_box, sizeof(struct wlr_box));
-		scale_box(&box, ctx->output->wlr_output->scale);
-		render_rect(ctx, &box, color);
-	}
+	wlr_scene_node_destroy(&e->indicator_rect->node);
+	e->indicator_rect = NULL;
```

Comment:
Do we need to free `e->indicator_rect` as well?

Reply:
You have a misunderstanding here. The node type is not allocated as a separate object. wlr_scene_node is the part of wlr_scene_rect/wlr_scene_tree whatever. Destroying the node will destroy the base type, along with all children of the node.